### PR TITLE
[SPARK-26021][SQL][followup] add test for special floating point values

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

a followup of https://github.com/apache/spark/pull/23043 . Add a test to show the minor behavior change introduced by #23043 , and add migration guide.

## How was this patch tested?

a new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @adoron @kiszk @viirya

**Comment by viirya** [comment]
retest this please.

File: docs/sql-migration-guide-upgrade.md [hunk 1]
```diff
@@ -17,14 +17,16 @@ displayTitle: Spark SQL Upgrading Guide
...
-  
+
   - In Spark version 2.4 and earlier, `Dataset.groupByKey` results to a grouped dataset with key attribute wrongly named as "value", if the key is non-struct type, e.g. int, string, array, etc. This is counterintuitive and makes the schema of aggregation queries weird. For example, the schema of `ds.groupByKey(...).count()` is `(value, count)`. Since Spark 3.0, we name the grouping attribute to "key". The old behaviour is preserved under a newly added configuration `spark.sql.legacy.dataset.nameNonStructGroupingKeyAsValue` with a default value of `false`.
 
+  - In Spark version 2.4 and earlier, float/double -0.0 is semantically equal to 0.0, but users can still distinguish them via `Dataset.show`, `Dataset.collect` etc. Since Spark 3.0, float/double -0.0 is replaced by 0.0 internally, and users can't distinguish them any more.
```
[the remark this answers is not in the document]
Comment (adoron) [reply docs/sql-migration-guide-upgrade.md]: What version of hive did you test? 
It was fixed in https://issues.apache.org/jira/browse/HIVE-11174

[the remark this answers is not in the document]
Comment (viirya) [reply docs/sql-migration-guide-upgrade.md]: I ran few simple queries on Hive 2.1.

Simple comparison seems ok:

```
hive> select 1 where 0.0=-0.0;
OK
1
Time taken: 0.047 seconds, Fetched: 1 row(s)
hive> select 1 where -0.0<0.0;
OK
Time taken: 0.053 seconds
```

But group by behavior seems not correct:
```
hive> select * from test;
OK
0.0
-0.0
0.0
Time taken: 0.11 seconds, Fetched: 3 row(s)
hive> select * from test;
OK
0.0
-0.0
0.0
Time taken: 0.11 seconds, Fetched: 3 row(s)
hive> select a, count(*) from test group by a;
-0.0	3
Time taken: 1.308 seconds, Fetched: 1 row(s)
```

**Review by srowen — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by kiszk** [comment]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

File: common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java [hunk 2]
```diff
@@ -165,10 +165,14 @@ public void writeMinusZeroIsReplacedWithZero() {
...
     Platform.putFloat(floatBytes, Platform.BYTE_ARRAY_OFFSET, -0.0f);
-    double doubleFromPlatform = Platform.getDouble(doubleBytes, Platform.BYTE_ARRAY_OFFSET);
-    float floatFromPlatform = Platform.getFloat(floatBytes, Platform.BYTE_ARRAY_OFFSET);
 
-    Assert.assertEquals(Double.doubleToLongBits(0.0d), Double.doubleToLongBits(doubleFromPlatform));
-    Assert.assertEquals(Float.floatToIntBits(0.0f), Float.floatToIntBits(floatFromPlatform));
+    byte[] doubleBytes2 = new byte[Double.BYTES];
+    byte[] floatBytes2 = new byte[Float.BYTES];
+    Platform.putDouble(doubleBytes, Platform.BYTE_ARRAY_OFFSET, 0.0d);
+    Platform.putFloat(floatBytes, Platform.BYTE_ARRAY_OFFSET, 0.0f);
```
Comment (dongjoon-hyun) [inline common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java]: Unfortunately, these should be `doubleByte2` and `floatBytes2`.
I'll comment on the follow-up PR https://github.com/apache/spark/pull/23239 , too. We can fix there.

Comment (cloud-fan) [reply common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java]: ah good catch! I'm surprised this test passed before...

**Pull request closed without merging**